### PR TITLE
Stop using std::map since cleanup races with the finalizer during shutdown.

### LINKED ISSUE #38531: Test failure: Interop\\COM\\NativeClients\\Events\\Events.cmd
failed in job: [ runtime-coreclr gcstress0x3-gcstress0xc 20200628.1 ](https://dev.azure.com/dnceng/public/_build/results?buildId=708505&view=ms.vss-test-web.build-test-results-tab&runId=21947472&resultId=103517&paneView=debug)

Error message
~~~

Return code: 1
Raw output file: C:\h\w\B0230973\w\A8450921\e\Interop\COM\Reports\Interop.COM\NativeClients\Events\Events.output.txt
Raw output:
BEGIN EXECUTION
Copying 'C:\h\w\B0230973\p\CoreShim.dll'...
 1 file(s) copied.
 "C:\h\w\B0230973\p\corerun.exe" Events.dll 
Searching for exe to launch in C:\h\w\B0230973\w\A8450921\e\Interop\COM\NativeClients\Events...
Launching 'C:\h\w\B0230973\w\A8450921\e\Interop\COM\NativeClients\Events\COMClientEvents.exe'...

Expected: 100
Actual: -1073741819
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=C:\h\w\B0230973\p
> C:\h\w\B0230973\w\A8450921\e\Interop\COM\NativeClients\Events\Events.cmd
Expected: True
Actual: False


Stack trace
   at Interop_COM._NativeClients_Events_Events_._NativeClients_Events_Events_cmd() in F:\workspace\_work\1\s\artifacts\tests\coreclr\Windows_NT.x64.Checked\TestWrappers\Interop.COM\Interop.COM.XUnitWrapper.cs:line 916
~~~

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
See https://github.com/dotnet/runtime/pull/38317#issuecomment-649197277

Fixes https://github.com/dotnet/runtime/issues/38531

/cc @AndyAyersMS @jkoritzinsky @elinor-fung

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
Wondering if this is sufficient, or the if the entire `EventSink` needs to be allocated somewhere other than the CRT heap?

**Comment by AaronRobinsonMSFT** [comment]
> Wondering if this is sufficient, or the if the entire `EventSink` needs to be allocated somewhere other than the CRT heap?

Boo. That does make sense. Not thrilled about changing that, but I can't argue with the logic.

File: src/coreclr/tests/src/Interop/COM/NativeClients/Events/EventTests.cpp [hunk 1]
```diff
@@ -16,28 +16,89 @@
 #define THROW_IF_FAILED(exp) { hr = exp; if (FAILED(hr)) { ::printf("FAILURE: 0x%08x = %s\n", hr, #exp); throw hr; } }
 #define THROW_FAIL_IF_FALSE(exp) { if (!(exp)) { ::printf("FALSE: %s\n", #exp); throw E_FAIL; } }
 
-#include <map>
```
Comment (elinor-fung) [inline src/coreclr/tests/src/Interop/COM/NativeClients/Events/EventTests.cpp:19]: Is it just the debug iterators causing issues? Can we set `_ITERATOR_DEBUG_LEVEL`?

Comment (AaronRobinsonMSFT) [reply src/coreclr/tests/src/Interop/COM/NativeClients/Events/EventTests.cpp:19]: Possibly. However @AndyAyersMS statement above about the state of the CRT being the real issue is why we may want to just remove its usage entirely.

**Comment by AaronRobinsonMSFT** [comment]
After looking at this again I think not updating the `new` alloc for the object is unlikely to impact this. I will of course reconsider if we observe a failure in practice, but then we will need to reconsider all native COM server tests.

I think the current issue addresses the far more complicated issue involving the way STL collections perform validation.

**Comment by AaronRobinsonMSFT** [comment]
Any other feedback @elinor-fung or @jkoritzinsky ?

**Review by elinor-fung — APPROVED** [review]

**Pull request merged**